Add support for request dataType to network data collection

juliandescottes opened the pull request:
Fixes #748 

- Adds another entry point expected to be called from the fetch spec `clone network request body` (alternatively could be called explicitly from beforeRequestSent for now)
- Extracts a generic algorithm `maybe collect network data` from `maybe collect network response body`
- Update `maybe collect network response body` to call `maybe collect network data`
- Adds a new algorithm `maybe collect network request body` called from the beforeRequestSent steps

Otherwise this mostly reuses the logic added for response body collection. Commands do not need to be udpated AFAICT


<!--
    This comment and the below content is programmatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://pr-preview.s3.amazonaws.com/juliandescottes/webdriver-bidi/pull/1011.html" title="Last updated on Sep 29, 2025, 9:30 AM UTC (2eca438)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/webdriver-bidi/1011/bf20c0b...juliandescottes:2eca438.html" title="Last updated on Sep 29, 2025, 9:30 AM UTC (2eca438)">Diff</a>

sadym-chromium commented on index.bs:
```diff
@@ -6876,11 +6876,40 @@ To <dfn>match collector for navigable</dfn> given |collector| and |navigable|:
 </div>
 
+<div algorithm>
+To <dfn export>clone network request body</dfn> given |request|:
```
oh, despute the comment, it is not used externally

sadym-chromium replied:
This comment is not a blocker (I already approved the PR). 

I cannot see in the [fetch RP](https://github.com/whatwg/fetch/pull/1540) any reference to this hook. Maybe this comment meant to underline that this is a part of the `WebDriver BiDi before request sent` hook, which is invoked by the fetch spec in [fetch RP](https://github.com/whatwg/fetch/pull/1540)?

juliandescottes replied:
Oh no you're right, I haven't updated James' PR. Not quite sure how to do that but I can give it a try.

sadym-chromium commented on index.bs:
```diff
@@ -6925,19 +6954,51 @@ To <dfn>maybe abort network response body collection</dfn> given |request|:
+</div>
+
+<div algorithm>
+To <dfn>maybe collect network data</dfn> given |request|, |collected data|, |size| and |data type|:
```
Even though we usually don't do it, we can provide types here for readability:
```suggestion
To <dfn>maybe collect network data</dfn> given [=/request=] |request|, [=network data=] |collected data|, js-uint |size| and [=network.DataType=] |data type|:
```

juliandescottes replied:
Done

juliandescottes commented on index.bs:
```diff
@@ -6876,11 +6876,40 @@ To <dfn>match collector for navigable</dfn> given |collector| and |navigable|:
+<div algorithm>
+To <dfn export>clone network request body</dfn> given |request|:
+
+Note: This hook is intended to be triggered by the fetch spec when the request body has been safely extracted.
+See step 9 of https://fetch.spec.whatwg.org/#concept-fetch
```
If that's about the fact the fetch spec doesn't include this, see my other comment about https://github.com/whatwg/fetch/pull/1540

And same proposal: we can trigger this from beforeRequestSent if we prefer?

Otherwise can you explain what is not valid?

sadym-chromium replied:
Resolving in favor of https://github.com/w3c/webdriver-bidi/pull/1011#discussion_r2387320976

juliandescottes commented:
Thanks for the review! I addressed the suggestions to add types. Let me know if you want to switch to an internal trigger for the algorithm for now. I'll discuss internally to make the fetch PR move.
